R-package/R/xgb.Booster.R

```diff
@@ -1,180 +1,88 @@
-# Construct an internal xgboost Booster and return a handle to it.
+# Construct an internal xgboost Booster and get its current number of rounds.
 # internal utility function
-xgb.Booster.handle <- function(params, cachelist, modelfile, handle) {
+# Note: the number of rounds in the C booster gets reset to zero when updating
+# the parameters through 'xgb.parameters', hence the need to return it from
+# this function when using it for training continuation
+xgb.Booster <- function(params, cachelist, modelfile, training_continuation) {
   if (typeof(cachelist) != "list" ||
       !all(vapply(cachelist, inherits, logical(1), what = 'xgb.DMatrix'))) {
     stop("cachelist must be a list of xgb.DMatrix objects")
   }
   ## Load existing model, dispatch for on disk model file and in memory buffer
   if (!is.null(modelfile)) {
-    if (typeof(modelfile) == "character") {
+    if (is.character(modelfile)) {
       ## A filename
-      handle <- .Call(XGBoosterCreate_R, cachelist)
+      bst <- .Call(XGBoosterCreate_R, cachelist)
       modelfile <- path.expand(modelfile)
-      .Call(XGBoosterLoadModel_R, handle, enc2utf8(modelfile[1]))
-      class(handle) <- "xgb.Booster.handle"
+      .Call(XGBoosterLoadModel_R, xgb.get.handle(bst), enc2utf8(modelfile[1]))
+      niter <- xgb.nrounds(bst)
       if (length(params) > 0) {
-        xgb.parameters(handle) <- params
+        xgb.parameters(bst) <- params
       }
-      return(handle)
-    } else if (typeof(modelfile) == "raw") {
+      return(list(bst = bst, niter = niter))
+    } else if (is.raw(modelfile)) {
       ## A memory buffer
-      bst <- xgb.unserialize(modelfile, handle)
+      bst <- xgb.unserialize(modelfile)
+      niter <- xgb.nrounds(bst)
       xgb.parameters(bst) <- params
-      return(bst)
+      return(list(bst = bst, niter = niter))
```

Comment:
If I understand correctly, you want to get this attribute on-demand and remove R attributes so that everything is consistent.

Reply:
Yes.

Reply:
Then is it necessary to cache it in a list?

Reply:
Yes, it's necessary to return it there, because in the case of updates, it gets reset to zero, but then there is a condition that checks:
```r
if (is_update && nrounds > niter_init)
    stop("nrounds cannot be larger than ", niter_init, " (nrounds of xgb_model)")
```

So, if one calls `XGBoosterBoostedRounds` after setting the parameters (which resets the rounds to zero in the case of updates), it will return zero and thus will not be possible to check that condition.

Or is there perhaps a better way of obtaining the current number of boosting iterations in order to check for that condition when performing an update?

Reply:
Since this function is a part of the public interface, I think we should think of a way to avoid unnecessary publicly accessible attributes. I don't have a good solution to the update process yet (has been like this for a while), but a quick workaround as an internal hack could be setting it as an internal booster attribute and deleting it after the check.

Reply:
Function `xgb.Booster` is not within the exported namespace, it's only meant for internal usage.

Reply:
thank you for sharing, then it's fine.